### PR TITLE
Adding new tests to libzim

### PR DESCRIPTION
**Pull request by Rishabhg71**
I have added the `-s DISABLE_EXCEPTION_CATCHING=0` to compile the files properly.

### CODE REVIEW

**Comment by Rishabhg71** [comment]
is it okay to let the GitHub actions run? I am seeing some steps where files may be updated on server

**Comment by Rishabhg71** [comment]
@Jaifroid Tests are now enabled on all the build files + could you answer my question above?

**Comment by Jaifroid** [comment]
@Rishabhg71 I'm sorry, I didn't see the notifications from your previous message and query. You can run GitHub actions. "Build and publish release artefacts" (https://github.com/openzim/javascript-libzim/actions/workflows/build_libzim_wasm.yml) explains what it does in the Run Workflow dropdown. I you run it as "nightly", then it will create artefacts and upload them to the nightly server. If you run it as "release" but without a version number, then it will just build artefacts and archive them under the workflow summary. If you add a version number, then it will create a draft release in releases (which can simply be deleted if it's a test run).

"Upload release assets to Kiwix" only runs after a draft release is published. So no need to worry about it if you're not actually pressing the "Publish" button on a draft release!

**Comment by Jaifroid** [comment]
Forgot to add that automatic running of workflows when changes to a PR are pushed only produce archived versions under the workflow, they don't publish them to the Kiwix server, so it's safe to let the workflow complete after push, or to re-run it.

**Comment by Rishabhg71** [comment]
> A difference between the production and the debug builds is that the latter have -s DISABLE_EXCEPTION_CATCHING=0 in the compile command. For more info, see https://emscripten.org/docs/porting/exceptions.html#emscripten-javascript-based-exception-support .
Rather than compiling the production versions with enabled exception catching, we should make use of https://github.com/emscripten-core/emscripten/blob/main/src/settings.js#L683 (EXCEPTION_CATCHING_ALLOWED = [ ]) to enable catching only in the affected function(s).

How do I know which functions should be passed as a parameter? I think `getEntryByPath` is the only function or am I wrong?

**Comment by Jaifroid** [comment]
@Rishabhg71 The functions for which we have written bindings are the articleCount, search, and getDirectoryByPath. We need to write some more, but they're not urgent, because we handle everything else in the backend ourselves currently,

**Comment by Rishabhg71** [comment]
@Jaifroid I have added `-s EXCEPTION_CATCHING_ALLOWED=["getEntryByPath","search","getArticleCount","loadArchive"]` but seems like I still that uncaught error. I am not sure why this is happening any idea? building with `-s DISABLE_EXCEPTION_CATCHING=0` is working fine
image

**Comment by Jaifroid** [comment]
@Rishabhg71 I'm afraid I don't know, and I've never used that setting allowing exception catching for individual functions. I would say that if the size of the bundle is not significantly increased by the general setting, it's probably safe to use it, unless you can find some info on how to get the narrower exception catching working...

**Comment by Rishabhg71** [comment]
I think the size is good enough it has only increased 20 kb max, files with suffix "1" are compiled by `DISABLE_EXCEPTION_CATCHING`
image

If this is acceptable I think then you can do a first review now

**Comment by Jaifroid** [comment]
OK, thanks, I'll review it, but in the meantime, you might want to revert the individual exception catching, which is still in the PR and is not working, right? Certainly the tests have failed, not 100% sure if it¡'s for that reason.

File: tests/prototype/chromium.e2e.runner.js [hunk 1]
```diff
@@ -17,4 +18,5 @@ async function loadChromiumDriver () {
     return driver;
 };
 
-rayCharles.runTests(await loadChromiumDriver());
+// rayCharles.runTests(await loadChromiumDriver());
+stackexchange.runTests(await loadChromiumDriver());
```
Comment (Jaifroid) [inline tests/prototype/chromium.e2e.runner.js:22]: I presume the Ray Charles test is temporarily disabled here, and you'll re-enable it later?

File: Makefile [hunk 2]
```diff
@@ -96,11 +96,11 @@ libzim-asm.dev.js: libzim_bindings.cpp prejs_file_api.js postjs_file_api.js
 
 # Production WASM version with WORKERFS and NODEFS, optimized and packed
 libzim-wasm.js: libzim_bindings.cpp prejs_file_api.js postjs_file_api.js
-	em++ -o libzim-wasm.js --bind libzim_bindings.cpp -I/src/build/include -L/src/build/lib -lzim -llzma -lzstd -lxapian -lz -licui18n -lpthread -licuuc -licudata -O3 --pre-js prejs_file_api.js --post-js postjs_file_api.js -s WASM=1 -s "EXPORTED_RUNTIME_METHODS=['ALLOC_NORMAL','err','ALLOC_STACK','out']" -s INITIAL_MEMORY=83886080 -s ALLOW_MEMORY_GROWTH=1 -s DYNAMIC_EXECUTION=0 -lworkerfs.js -lnodefs.js
+	em++ -o libzim-wasm.js --bind libzim_bindings.cpp -I/src/build/include -L/src/build/lib -lzim -llzma -lzstd -lxapian -lz -licui18n -lpthread -licuuc -licudata -O3 --pre-js prejs_file_api.js --post-js postjs_file_api.js -s WASM=1 -s "EXPORTED_RUNTIME_METHODS=['ALLOC_NORMAL','err','ALLOC_STACK','out']" -s INITIAL_MEMORY=83886080 -s EXCEPTION_CATCHING_ALLOWED=["getEntryByPath","search","getArticleCount","loadArchive"] -s ALLOW_MEMORY_GROWTH=1 -s DYNAMIC_EXECUTION=0 -lworkerfs.js -lnodefs.js
```
Comment (Jaifroid) [inline Makefile:100]: I presume this type of exception catching need to be reverted to the catch-all type.

File: Makefile [hunk 3]
```diff
@@ -96,11 +96,11 @@ libzim-asm.dev.js: libzim_bindings.cpp prejs_file_api.js postjs_file_api.js
...
 
 # Production ASM version with WORKERFS and NODEFS, optimized and packed
 libzim-asm.js: libzim_bindings.cpp prejs_file_api.js postjs_file_api.js
-	em++ -o libzim-asm.js --bind libzim_bindings.cpp -I/src/build/include -L/src/build/lib -lzim -llzma -lzstd -lxapian -lz -licui18n -licuuc -licudata -O3 --pre-js prejs_file_api.js --post-js postjs_file_api.js -s WASM=0 --memory-init-file 0 -s MIN_EDGE_VERSION=40 -s "EXPORTED_RUNTIME_METHODS=['ALLOC_NORMAL','err','ALLOC_STACK','out']" -s INITIAL_MEMORY=83886080 -s ALLOW_MEMORY_GROWTH=1 -s DYNAMIC_EXECUTION=0 -lworkerfs.js -lnodefs.js
+	em++ -o libzim-asm.js --bind libzim_bindings.cpp -I/src/build/include -L/src/build/lib -lzim -llzma -lzstd -lxapian -lz -licui18n -licuuc -licudata -O3 --pre-js prejs_file_api.js --post-js postjs_file_api.js -s WASM=0 --memory-init-file 0 -s MIN_EDGE_VERSION=40 -s "EXPORTED_RUNTIME_METHODS=['ALLOC_NORMAL','err','ALLOC_STACK','out']" -s EXCEPTION_CATCHING_ALLOWED=["getEntryByPath","search","getArticleCount","loadArchive"] -s INITIAL_MEMORY=83886080 -s ALLOW_MEMORY_GROWTH=1 -s DYNAMIC_EXECUTION=0 -lworkerfs.js -lnodefs.js
```
Comment (Jaifroid) [inline Makefile:104]: Ditto

**Review by Jaifroid — COMMENTED** [review]
It's all looking good to me -- well written code! Just the few queries below, and I'd like to see a run passing with both Ray Charles and Stackexchange being tested.

**Comment by Rishabhg71** [comment]
@Jaifroid Sorry I forgot to revert my changes, you can check again

File: tests/prototype/stackexchange.e2e.spec.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+ *  Run the tests
+ * @param {WebDriver} driver Selenium WebDriver object
+ * @returns {Promise<void>}  A Promise for the completion of the tests
+*/
+function runTests (driver) {
+    let browserName, browserVersion;
+    driver.getCapabilities().then(function (caps) {
+        browserName = caps.get('browserName');
+        browserVersion = caps.get('browserVersion');
+        console.log('\nRunning Legacy Ray Charles tests on: ' + browserName + ' ' + browserVersion);
```
Comment (Jaifroid) [inline tests/prototype/stackexchange.e2e.spec.js:24]: I think this should says "Running StackExchange tests on..."

File: tests/prototype/ray_charles.e2e.spec.js [hunk 5]
```diff
@@ -25,67 +27,78 @@ function runTests (driver) {
     // Set the implicit wait to 3 seconds
```
Comment (Jaifroid) [inline tests/prototype/ray_charles.e2e.spec.js:27]: I can't add a comment in the right place here, but on line 24 of this file, place change "Legacy Ray Charles" to "Ray Charles". Ths ZIM here is a modern ZIM, not a legacy ZIM. (This was my mistake, I know you didn't change that code.)

**Review by Jaifroid — APPROVED** [review]
Looks very good to me, with just the two small changes below! After this is merged, we should create a new release, so that we can use officially released artefacts over at Kiwix JS.

**Pull request merged**